README
NAME

    WebService::Client - A base role for quickly and easily creating web
    service clients

VERSION

    version 0.0203

SYNOPSIS

        {
            package WebService::Foo;
            use Moo;
            with 'WebService::Client';
    
            use Function::Parameters;
    
            has '+base_url' => ( default => 'https://foo.com/v1' );
            has auth_token  => ( is => 'ro', required => 1 );
    
            method BUILD() {
                $self->ua->default_header('X-Auth-Token' => $self->auth_token);
                # or if the web service uses http basic/digest authentication:
                # $self->ua->credentials( ... );
                # or
                # $self->ua->default_headers->authorization_basic( ... );
            }
    
            method get_widgets() {
                return $self->get("/widgets");
            }
    
            method get_widget($id) {
                return $self->get("/widgets/$id");
            }
    
            method create_widget($widget_data) {
                return $self->post("/widgets", $widget_data);
            }
        }
    
        my $client = WebService::Foo->new(
            auth_token => 'abc',
            logger     => Log::Tiny->new('/tmp/foo.log'), # optional
            timeout    => 10, # optional, defaults to 10
            retries    => 0,  # optional, defaults to 0
        );
        my $widget = $client->create_widget({ color => 'blue' });
        print $client->get_widget($widget->{id})->{color};

DESCRIPTION

    This module is a base role for quickly and easily creating web service
    clients. Every time I created a web service client, I noticed that I
    kept rewriting the same boilerplate code independent of the web
    service. This module does the boring boilerplate for you so you can
    just focus on the fun part - writing the web service specific code.

METHODS

    These are the methods this role composes into your class. The HTTP
    methods (get, post, put, and delete) will return the deserialized
    response data, assuming the response body contained any data. This will
    usually be a hashref. If the web service responds with a failure, then
    the corresponding HTTP response object is thrown as an exception. This
    exception is a HTTP::Response object that has the
    HTTP::Response::Stringable role so it can be stringified. GET requests
    that result in 404 or 410 will not result in an exception. Instead,
    they will simply return undef.

    The `get/post/put/delete` methods all can take an optional headers
    keyword argument that is a hashref of custom headers.

 get

        $client->get('/foo');
        $client->get('/foo', headers => { foo => 'bar' });

    Makes an HTTP POST request.

 post

        $client->post('/foo', { some => 'data' });
        $client->post('/foo', { some => 'data' }, headers => { foo => 'bar' });

    Makes an HTTP POST request.

 put

        $client->put('/foo', { some => 'data' });

    Makes an HTTP PUT request.

 delete

        $client->delete('/foo');

    Makes an HTTP DELETE request.

 req

        my $req = HTTP::Request->new(...);
        $client->req($req);

    This is called internally by the above HTTP methods. You will usually
    not need to call this explicitly. It is exposed as part of the public
    interface in case you may want to add a method modifier to it. Here is
    a contrived example:

        around req => sub {
            my ($orig, $self, $req) = @_;
            $req->authorization_basic($self->login, $self->password);
            return $self->$orig($req, @rest);
        };

 log

    Logs a message using the provided logger.

ATTRIBUTES

 base_url

    This is the only attribute that is required. This is the base url that
    all request will be made against.

 ua

    Optional. A proper default LWP::UserAgent will be created for you.

 timeout

    Optional. Default is 10.

 retries

    Optional. Default is 0.

 logger

    Optional.

 content_type

    Optional. Default is 'application/json'.

EXAMPLES

    Here are some examples of web service clients built with this role. You
    can view their source to help you get started.

      * Business::BalancedPayments

      * WebService::HipChat

      * WebService::Lob

      * WebService::SmartyStreets

      * WebService::Stripe

SEE ALSO

      * Net::HTTP::API

      * Role::REST::Client

AUTHOR

    Naveed Massjouni <naveed@vt.edu>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Naveed Massjouni.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
==================================================
Changes from 2004-12-04 00:00:00 +0000 to present.
==================================================

-------------------------------------------
version 0.0203 at 2014-12-02 07:31:40 +0000
-------------------------------------------

  Change: 5c4a24ab100a29ad7475c9efe793cab1c398311f
  Author: Naveed Massjouni <naveed@vt.edu>
  Date : 2014-12-02 02:29:47 +0000

    v0.0203 better content-type detection 

  Change: 22c14aa437de0ce5d21e71d59ac39e23028a5e2e
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-19 03:33:02 +0000

    added WebService::Stripe to examples 

-------------------------------------------
version 0.0202 at 2014-11-19 06:32:02 +0000
-------------------------------------------

  Change: 725a7d85d0136e9783785285d98cbec184de66f3
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-19 01:32:02 +0000

    v0.0202 

  Change: 3973e888ca21255b127d16beea31b4d6a6fff72a
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-19 01:30:47 +0000

    better logging and require latest HTTP::Message 

  Change: 62e07ff76045154b85d775f2267cfa83c75c4f79
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-18 10:26:53 +0000

    v0.0201 

  Change: 453481aa994390f8f52cd07c74d4cb086941cbf0
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-18 10:22:58 +0000

    don't squash custom Content-Type header if it is passed in 

  Change: eae81021cbe77207897a8c9d7b42980a25855e6c
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-18 07:59:56 +0000

    allow headers as hashref or arrayref 

  Change: 08e5f209b43fde2bcbdcfa4f38dd6efdee3a77b8
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 19:03:06 +0000

    require HTTP::Response::Stringable v0.0002 

  Change: 4898e4ebcfa2095f45ea719d2fd939c77b969f4a
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 18:55:22 +0000

    Merge pull request #4 from ironcamel/custom-headers

    allow custom headers to be provided 

-------------------------------------------
version 0.0200 at 2014-11-17 23:54:11 +0000
-------------------------------------------

  Change: be6ece67554b62ed05ae1f64c61242008551ba2f
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 18:54:11 +0000

    v0.0200 

  Change: 4199d427ad05c705baec939cb734d18bdd6ce8fd
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 18:50:32 +0000

    use HTTP::Response and HTTP::Request's built in as_string() for
    logging 

  Change: 63541d4b92c0325754ead75ab2b8de192f28753e
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 17:48:36 +0000

    allow custom headers to be provided 

-------------------------------------------
version 0.0100 at 2014-11-12 22:03:20 +0000
-------------------------------------------

  Change: 7f09133c780ab68789739655e646933fa622f901
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-12 17:03:20 +0000

    more docs 

  Change: d28a38f4722e8b31289106e7e480f2d3ce228b52
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-12 16:51:36 +0000

    Merge pull request #2 from ironcamel/content-type

    added support for different request content types 

  Change: be423db958ee19ad6130eb133d3ce3002d79cea5
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-12 16:41:15 +0000

    added support for different request content types 

  Change: f2a672f19f4a9f486febca710466d2878e14348a
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-09-26 02:39:39 +0000

    initial commit 

================
End of releases.
================


lib/WebService/Client.pm
package WebService::Client;
use Moo::Role;

our $VERSION = '0.0203'; # VERSION

use Carp qw(croak);
use HTTP::Request::Common qw(DELETE GET POST PUT);
use JSON qw(decode_json encode_json);
use LWP::UserAgent;

has base_url => ( is => 'ro', required => 1 );

has ua => (
    is      => 'ro',
    lazy    => 1,
    default => sub {
        my $self = shift;
        my $ua = LWP::UserAgent->new;
        $ua->timeout($self->timeout);
        return $ua;
    },
);

has timeout => ( is => 'ro', default => 10 );

has retries => ( is => 'ro', default => 0 );

has logger => ( is => 'ro' );

has content_type => (
    is      => 'rw',
    default => 'application/json',
);

sub get {
    my ($self, $path, $params, %args) = @_;
    $params ||= {};
    my $headers = $self->_headers(%args);
    my $url = $self->_url($path);
    my $q = '';
    if (%$params) {
        $q = '?' . join '&', map { "$_=$params->{$_}" } keys %$params;
    }
    return $self->req(GET "$url$q", %$headers);
}

sub post {
    my ($self, $path, $params, %args) = @_;
    my $headers = $self->_headers(%args);
    my $url = $self->_url($path);
    return $self->req(POST $url, %$headers, _content($params, $headers));
}

sub put {
    my ($self, $path, $params, %args) = @_;
    my $headers = $self->_headers(%args);
    my $url = $self->_url($path);
    return $self->req(PUT $url, %$headers, _content($params, $headers));
}

sub delete {
    my ($self, $path, %args) = @_;
    my $headers = $self->_headers(%args);
    my $url = $self->_url($path);
    return $self->req(DELETE $url, %$headers);
}

sub req {
    my ($self, $req) = @_;
    $self->_log_request($req);
    my $res = $self->ua->request($req);
    Moo::Role->apply_roles_to_object($res, 'HTTP::Response::Stringable');
    $self->_log_response($res);

    my $retries = $self->retries;
    while ($res->code =~ /^5/ and $retries--) {
        sleep 1;
        $res = $self->ua->request($req);
        $self->_log_response($res);
    }

    return undef if $req->method eq 'GET' and $res->code =~ /404|410/;
    die $res unless $res->is_success;
    return $res->content ? decode_json($res->content) : 1;
}

sub log {
    my ($self, $msg) = @_;
    return unless $self->logger;
    $self->logger->DEBUG($msg);
}

sub _url {
    my ($self, $path) = @_;
    croak 'The path is missing' unless defined $path;
    return $path =~ /^http/ ? $path : $self->base_url . $path;
}

sub _headers {
    my ($self, %args) = @_;
    my $headers = $args{headers} || {};
    croak 'The headers param must be a hashref' unless 'HASH' eq ref $headers;
    $headers->{content_type} = $self->content_type
        unless _content_type($headers);
    return $headers;
}

sub _log_request {
    my ($self, $req) = @_;
    $self->log(ref($self) . " REQUEST:\n" . $req->as_string);
}

sub _log_response {
    my ($self, $res) = @_;
    $self->log(ref($self) . " RESPONSE:\n" . $res->as_string);
}

sub _content_type {
    my ($headers) = @_;
    return $headers->{'Content-Type'}
        || $headers->{'content-type'}
        || $headers->{content_type};
}

sub _content {
    my ($params, $headers) = @_;
    my @content;
    if (defined $params) {
        $params = encode_json $params if _content_type($headers) =~ /json/;
        @content = ( content => $params );
    }
    return @content;
}

# ABSTRACT: A base role for quickly and easily creating web service clients


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

WebService::Client - A base role for quickly and easily creating web service clients

=head1 VERSION

version 0.0203

=head1 SYNOPSIS

    {
        package WebService::Foo;
        use Moo;
        with 'WebService::Client';

        use Function::Parameters;

        has '+base_url' => ( default => 'https://foo.com/v1' );
        has auth_token  => ( is => 'ro', required => 1 );

        method BUILD() {
            $self->ua->default_header('X-Auth-Token' => $self->auth_token);
            # or if the web service uses http basic/digest authentication:
            # $self->ua->credentials( ... );
            # or
            # $self->ua->default_headers->authorization_basic( ... );
        }

        method get_widgets() {
            return $self->get("/widgets");
        }

        method get_widget($id) {
            return $self->get("/widgets/$id");
        }

        method create_widget($widget_data) {
            return $self->post("/widgets", $widget_data);
        }
    }

    my $client = WebService::Foo->new(
        auth_token => 'abc',
        logger     => Log::Tiny->new('/tmp/foo.log'), # optional
        timeout    => 10, # optional, defaults to 10
        retries    => 0,  # optional, defaults to 0
    );
    my $widget = $client->create_widget({ color => 'blue' });
    print $client->get_widget($widget->{id})->{color};

=head1 DESCRIPTION

This module is a base role for quickly and easily creating web service clients.
Every time I created a web service client, I noticed that I kept rewriting the
same boilerplate code independent of the web service.
This module does the boring boilerplate for you so you can just focus on
the fun part - writing the web service specific code.

=head1 METHODS

These are the methods this role composes into your class.
The HTTP methods (get, post, put, and delete) will return the deserialized
response data, assuming the response body contained any data.
This will usually be a hashref.
If the web service responds with a failure, then the corresponding HTTP
response object is thrown as an exception.
This exception is a L<HTTP::Response> object that has the
L<HTTP::Response::Stringable> role so it can be stringified.
GET requests that result in 404 or 410 will not result in an exception.
Instead, they will simply return C<undef>.

The `get/post/put/delete` methods all can take an optional headers keyword
argument that is a hashref of custom headers.

=head2 get

    $client->get('/foo');
    $client->get('/foo', headers => { foo => 'bar' });

Makes an HTTP POST request.

=head2 post

    $client->post('/foo', { some => 'data' });
    $client->post('/foo', { some => 'data' }, headers => { foo => 'bar' });

Makes an HTTP POST request.

=head2 put

    $client->put('/foo', { some => 'data' });

Makes an HTTP PUT request.

=head2 delete

    $client->delete('/foo');

Makes an HTTP DELETE request.

=head2 req

    my $req = HTTP::Request->new(...);
    $client->req($req);

This is called internally by the above HTTP methods.
You will usually not need to call this explicitly.
It is exposed as part of the public interface in case you may want to add
a method modifier to it.
Here is a contrived example:

    around req => sub {
        my ($orig, $self, $req) = @_;
        $req->authorization_basic($self->login, $self->password);
        return $self->$orig($req, @rest);
    };

=head2 log

Logs a message using the provided logger.

=head1 ATTRIBUTES

=head2 base_url

This is the only attribute that is required.
This is the base url that all request will be made against.

=head2 ua

Optional. A proper default LWP::UserAgent will be created for you.

=head2 timeout

Optional.
Default is 10.

=head2 retries

Optional.
Default is 0.

=head2 logger

Optional.

=head2 content_type

Optional.
Default is C<'application/json'>.

=head1 EXAMPLES

Here are some examples of web service clients built with this role.
You can view their source to help you get started.

=over

=item *

L<Business::BalancedPayments>

=item *

L<WebService::HipChat>

=item *

L<WebService::Lob>

=item *

L<WebService::SmartyStreets>

=item *

L<WebService::Stripe>

=back

=head1 SEE ALSO

=over

=item *

L<Net::HTTP::API>

=item *

L<Role::REST::Client>

=back

=head1 AUTHOR

Naveed Massjouni <naveed@vt.edu>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Naveed Massjouni.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
